pkg/agent/metrics/prometheus.go

```diff
@@ -0,0 +1,103 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package metrics
+
+import (
+	"strconv"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"k8s.io/klog"
+
+	"github.com/vmware-tanzu/antrea/pkg/agent/interfacestore"
+	"github.com/vmware-tanzu/antrea/pkg/agent/openflow"
+	"github.com/vmware-tanzu/antrea/pkg/util/env"
+)
+
+type ovsStatManager struct {
+	ofClient     openflow.Client
+	ovsBridge    string
+	ovsTableDesc *prometheus.Desc
+}
+
+func (c *ovsStatManager) getOVSStatistics() (ovsFlowsByTable map[string]float64) {
+	ovsFlowsByTable = make(map[string]float64)
+	flowTableStatus := c.ofClient.GetFlowTableStatus()
+	for _, tableStatus := range flowTableStatus {
+		ovsFlowsByTable[strconv.Itoa(int(tableStatus.ID))] = float64(tableStatus.FlowCount)
+	}
+	return
+}
+
+func (c *ovsStatManager) Describe(ch chan<- *prometheus.Desc) {
+	ch <- c.ovsTableDesc
+}
+
+func (c *ovsStatManager) Collect(ch chan<- prometheus.Metric) {
+	ovsFlowsByTable := c.getOVSStatistics()
+	for tableId, tableFlowCount := range ovsFlowsByTable {
+		ch <- prometheus.MustNewConstMetric(
+			c.ovsTableDesc,
+			prometheus.GaugeValue,
+			tableFlowCount,
+			tableId,
+		)
+	}
+}
+
+func newOVSStatManager(ovsBridge string, ofClient openflow.Client) *ovsStatManager {
+	return &ovsStatManager{
+		ofClient:  ofClient,
+		ovsBridge: ovsBridge,
+		ovsTableDesc: prometheus.NewDesc(
+			"antrea_agent_ovs_flow_table",
+			"OVS flow table flow count.",
+			[]string{"table_id"},
+			prometheus.Labels{"bridge": ovsBridge},
+		),
+	}
+}
+
+func InitializePrometheusMetrics(
+	ovsBridge string,
+	ifaceStore interfacestore.InterfaceStore,
+	ofClient openflow.Client) {
+
+	klog.Info("Initializing prometheus metrics")
+	if err := prometheus.Register(prometheus.NewGaugeFunc(
+		prometheus.GaugeOpts{
+			Name: "antrea_agent_local_pod_count",
+			Help: "Number of pods on local node which are managed by the Antrea Agent.",
+		},
+		func() float64 { return float64(ifaceStore.GetContainerInterfaceNum()) },
+	)); err != nil {
+		klog.Error("Failed to register antrea_agent_local_pod_count with Prometheus")
+	}
+
+	nodeName, err := env.GetNodeName()
+	if err != nil {
+		klog.Errorf("Failed to retrieve agent K8S node name: %v", err)
+	}
+
+	gaugeHost := prometheus.NewGauge(prometheus.GaugeOpts{
+		Name:        "antrea_agent_runtime_info",
+		Help:        "Antrea agent runtime info , defined as a labels. The value of the gauge is always set to 1.",
+		ConstLabels: prometheus.Labels{"k8s_nodename": nodeName, "k8s_podname": env.GetPodName()},
+	})
+	gaugeHost.Set(1)
+	prometheus.MustRegister(gaugeHost)
```

Comment:
maybe a stupid question, but why do we use `MustRegister` here and `Register` for `"antrea_agent_local_pod_count"`?

Reply:
would like to know this as well :) what's the difference ?

Reply:
The reason to the difference is that I used different sample code for both :)
MustRegister() will panic on failure while Register() returns an error.
I'm guessing that using Register() is the correct approach here unless you think otherwise.

Reply:
Seems like I don't even have to register: it'll register with the default registry, I guess. I'll remove register calls.

Reply:
Sounds good